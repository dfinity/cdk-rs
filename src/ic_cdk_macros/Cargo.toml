[package]
name = "ic_cdk_macros"
version = "0.1.0"
authors = ["Hans Larsen <hans@larsen.online>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
candid = { git = "https://github.com/hansl/candid", branch = "rust-cdk" }
syn = { version = "1.0", features = ["fold", "full"] }
quote = "1.0"
proc-macro2 = "1.0"
