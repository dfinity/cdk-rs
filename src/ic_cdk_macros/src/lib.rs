use proc_macro::TokenStream;
use std::sync::atomic::{AtomicU32, Ordering};

mod error;
mod export;
mod import;

static NEXT_ID: AtomicU32 = AtomicU32::new(0);
pub(crate) fn id() -> u32 {
    NEXT_ID.fetch_add(1, Ordering::SeqCst)
}

fn handle_debug_and_errors<F>(
    cb: F,
    name: &str,
    attr: TokenStream,
    item: TokenStream,
) -> TokenStream
where
    F: FnOnce(TokenStream, TokenStream) -> Result<TokenStream, error::Errors>,
{
    if std::env::var_os("IC_CDK_DEBUG").is_some() {
        eprintln!("--- IC_CDK_MACROS DEBUG ---");
        eprintln!("{}\n    attr: {}\n    item: {}", name, attr, item);
    }

    let result = cb(attr, item);

    if std::env::var_os("IC_CDK_DEBUG").is_some() {
        eprintln!("--------- RESULT  ---------");
        if let Ok(ref stream) = result {
            eprintln!("{}", stream);
        }
        eprintln!("---------------------------");
    }

    // We return an empty tokenstream on a fatal error as not to display the error
    // twice.
    let (errors, res) = result.map_or_else(
        |e| (e, TokenStream::new()),
        |v| (error::Errors::default(), v),
    );

    errors.emit();
    res
}

#[proc_macro_attribute]
pub fn query(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_query, "ic_query", attr, item)
}

#[proc_macro_attribute]
pub fn update(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_update, "ic_update", attr, item)
}

#[proc_macro_attribute]
pub fn init(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_init, "ic_init", attr, item)
}

#[proc_macro_attribute]
pub fn import(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(import::ic_import, "ic_import", attr, item)
}
