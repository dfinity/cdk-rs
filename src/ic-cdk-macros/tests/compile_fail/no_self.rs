use ic_cdk_macros::query;

#[query]
fn method(self) {}

fn main() {}
