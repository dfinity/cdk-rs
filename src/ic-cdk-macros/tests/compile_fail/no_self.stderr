error: #[query] cannot be above functions with `self` as a parameter.
 --> tests/compile_fail/no_self.rs:4:11
  |
4 | fn method(self) {}
  |           ^^^^
